README
1. Introduction

itseml is a Python module that simplifies the setup of an emulated ITS station
by setting up a network namespace, and starting the various components as
services after the configuration for each component has been applied.

2. Usage

Starting/stopping stations happens dynamically. Specifically, the ws-itseml
and cli-itseml expect to receive an appropriately formatted JSON over
WebSockets and stdin respectively.

A sample JSON to start a station:
{
  "action": "start",
  "id": 1,               # Must be unique. It is used to identify the netns
  "stationid": 1234      # StationID used in the mw-server configuration.
  "gn": {
     "gn_addr": "0000:aabb:ccdd:eeff" # GeoNetworking Address
     "geobc_fwd_alg": 0  # Set to 4 to disable GeoBroadcast forwarding
  },
  "denm": {
     "forwarding": true  # Set to false to disable DENM forwarding
     "forceactionid": false
     "autoupdate": false
  },
  "position": {          # Define static position of the station
     "lat": "48.837590",
     "lon": "2.101363"
  }
}

The 'gn', 'stationid', and 'denm' objets are optional.
If 'gn' is not set, a random GN address is used, and the default GeoBC
forwarding algorithm is used.
If 'stationid' is not set, a random ID is generated for the station.
If 'denm' is not set, the values displayed above are used.

To stop the station:
{
  "action": "stop",
  "id": 1
}

Any additional fields will be ignored.

cli-itseml is provided as a testing aid. The usage is: cli-itseml < file.json
ws-itseml listens on port 8080, although in practice it sits behind a proxy.

The provided nginx configuration (doc/sample.nginx.conf) is set to proxy
two locations:
 - /ctl
 - /env/<envid>/pubsub/pub/<msgid>/<appid>

The first proxies requests to ws-itseml.
The second is used to communication with the middleware WebSocket API of each
station. To achieve this, nginx must be compiled with an embedded Lua
interpreter (on debian, use the nginx-extras package).

3. Installation

First, build the deb package with:
        make

The package will install:
 - the python module
 - the cli-itseml and ws-itseml frontends (in /usr/bin)
 - the required systemd unit files (in /lib/systemd/system)
 - some documentation in /usr/share/doc/itseml/

For a fully functional setup, you need to:
 - configure the v2xbr bridge (see below)
 - install and configure nginx:
    * apt-get install nginx-extras # We need Lua support
    * cp /usr/share/doc/itseml/sample.nginx.conf /etc/nginx/sites-available
    * ln -sf /etc/nginx/sites-available/sample.nginx.conf /etc/nginx/sites-enabled/default

The last command will only overwrite the symlink; the default file is left
untouched. Adjust if necessary.

4. Internals

An ITS stack is currently defined as:
 - itsnet
 - mwtun
 - mw-server
 - gpsd

Each ITS stack runs in a network namespace named:
 - envX

For each instance, two veth pairs are created:
 - vtxX <-> veth0
 - ctlX <-> ctl0

The different vtxX devices are bridged together, in a bridge called v2xbr.

There are various components/configurations needed to achieve a workable
setup:

 - bridge:
  On generic Linux systems:
        ip link add v2xbr type bridge # or "brctl addbr v2xbr"
        ip link set v2xbr up

  On Debian-like systems, the appropriate stanza for /etc/network/interfaces
is:
        auto v2xbr
        iface v2xbr inet manual
            bridge_ports none
            bridge_fd 0
            bridge_maxwait 0
            bridge_stp off

 - netns:
        The network namespace is setup via the appropriate script (installed
  as /usr/bin/setup_netns). The script is called by the eml-netns@.service unit.

 - itsnet:
        The GeoNetworking daemon is started by the eml-itsnet@.service unit.

 - mwtun:
        Interface betwen itsnet and mw-server. Started by the
eml-mwtun@service unit.

 - gps:
        The gpsd daemon fetches the position from the middleware. Since the
  middleware does not serve data directly over TCP, a proxy is used. The
  relevant units are:
           * eml-gpsd@.service
           * eml-gpsfwd@.service
           * eml-gpspipe@.service

        The eml-gpspipe service is simply a gpspipe process that runs for 3
  seconds and then quits before being restarted again by systemd. It is not
  needed per se but it helps mitigate a race condition:
          - gpsd is started
          - itsnet is started, attempts to connect to gpsd
          - upon receiving the connection, gpsd "activates" the upstream, i.e.
            the connection to the proxy.
          - the proxy isn't initialized yet, so gpsd fails.

        At this point, gpsd keeps the client connections open, but does not
  retry the upstream. In other words, itsnet is connected but does not receive
  any data.  Attempts to reconnect only happen when a new client connects,
  which is the reason why the gpspipe service exists.

 - mw-server:
        The middleware server is started the eml-mw-server@.service unit.

 - ws-itseml:
        A simple Python wrapper using gevent-websocket to convert received
  JSON data to a dict, and feed it to the itseml module.


debian/changelog
itseml (0.1-1) unstable; urgency=low

  * source package automatically created by stdeb 0.8.2

 -- Marios Makassikis <marios.makassikis@yogoko.fr>  Mon, 08 Feb 2016 21:18:56 +0100


itseml/itseml.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
import collections
import distutils.dir_util
import fcntl
import logging
import os
import os.path
import socket
import struct
import subprocess
import sys
import pprint
import random
import netaddr

from shutil import copyfile
from string import Template

TMPL_PATH = os.path.join(os.path.dirname(__file__), "templates")
CONF_PATH = "/var/run/itseml/"
IP_NETWORK= "10.1.1.0/24"
SUBNET_PLEN = 30

logger = logging.getLogger()
logger.setLevel(logging.DEBUG)

# create formatter
formatter = logging.Formatter('%(asctime)s :: %(levelname)s :: %(message)s')

# log to stderr
stream_handler = logging.StreamHandler()
stream_handler.setLevel(logging.DEBUG)
stream_handler.setFormatter(formatter)

logger.addHandler(stream_handler)

def start_env(params, envname):
    directory = "/var/run/itseml/%s/mw/config" % (envname)
    logging.info("Creating destination directory: %s", directory)

    try:
        distutils.dir_util.mkpath(directory)
    except DistutilsFileError as e:
        logging.error("Failed to create directory: %s" % (e))

    generate_configuration(params, envname)

    _service_action('start', envname)
    _network_conf(params['id'])
    logging.info("Started environment: %s" % (envname))
    return "200 OK"

def _mac_to_gn_addr(hw_addr):
    gn_addr = '0000' + hw_addr.replace(':', '')
    return ':'.join([gn_addr[i:i+4] for i in range(0, len(gn_addr), 4)])

def _get_if_mac(iface):
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    info = fcntl.ioctl(s.fileno(), 0x8927,  struct.pack('256s', iface[:15]))
    return ':'.join(['%02x' % ord(char) for char in info[18:24]])

def _network_conf(envnum):
    network = netaddr.IPNetwork(IP_NETWORK)
    subnets = list(network.subnet(SUBNET_PLEN))
    net = subnets[envnum-1]

    local = str(netaddr.IPAddress(net.first+1))
    remt = str(netaddr.IPAddress(net.first+2))

    cmds = []
    cmds.append("ip netns exec env%d ip a a %s/%d dev eth1" % (envnum, remt, SUBNET_PLEN))
    cmds.append("ip netns exec env%d ip r a default via %s" % (envnum, local))
    cmds.append("ip a a %s/%d dev ctl%d" % (local, SUBNET_PLEN, envnum))
    logging.info("Applying IP configuration")
    for cmd in cmds:
        out = subprocess.check_call(cmd, shell=True)

def _service_action(action, envname):
    svcs = ['eml-netns', 'eml-itsnet', 'eml-mwtun', 'eml-mw-server',
            'eml-gpsfwd', 'eml-gpsd', 'eml-gpspipe']
    if "is-active" in action:
        svcs.remove('eml-gpspipe')

    services = ' '.join([x + '@%s' % (envname) for x in svcs])

    subprocess.check_call("systemctl %s %s" % (action, services), shell=True)

def stop_env(envname, envid):
    _service_action('stop', envname)
    logging.info("Stopped environment: %s" % (envname))
    return "200 OK"

def status_env(envname):
    _service_action('is-active -q', envname)

def generate_configuration(params, envname):
    # Copy files that don't need modifications
    filelist = ['log4j.properties', 'config/spatconfig.xml', 'config/trafficlight.xml',
            'config/v2xconfig.xml', 'config/vehiclediagnosticconfig.xml', 'config/ldmservice.xml',
            'config/rhsservice.xml', 'config/dsrcmitigation.xml']

    src = [os.path.join(TMPL_PATH, "mw", x) for x in filelist]
    dst = [os.path.join("/var/run/itseml/%s/mw" % (envname), x) for x in filelist]
    for i,_ in enumerate(src):
        copyfile(src[i], dst[i])

    # Generate configuration files from JSON
    _fields = {
        'gn': 'itsnet.conf',
        'denm': 'mw/config/denservice.xml',
        'position': 'mw/config/positionproviderconfig.xml',
        'cam': 'mw/config/caconfig.xml',
    }

    def _process(field, filename):
        # Replace True with "true" and False with "false"
        for k, v in field.iteritems():
            if type(field[k]) == bool:
                field[k] = str(v).lower()

        with open(os.path.join(TMPL_PATH, filename), 'r') as f, \
            open(os.path.join(CONF_PATH, envname, filename), 'w') as dst:
            tpl = Template(f.read())
            dst.write(tpl.substitute(field))

    for k, v in _fields.iteritems():
        if k in params:
            logging.info("Creating configuration for %s: %s", k, v)
            _process(params[k], v)

    # Generate choirconf.xml
    tpl_params = {
        "station_id": params.get("stationid"),
        "station_type": 15,
    }
    logging.info("Creating configuration for mw-server: choirconf.xml")
    _process(tpl_params, "mw/choirconf.xml")

def _dict_update(source, overrides):
    """Update a nested dictionary

    Modify ``source`` in place.
    """
    for key, value in overrides.iteritems():
        if isinstance(value, collections.Mapping) and value:
            returned = _dict_update(source.get(key, {}), value)
            source[key] = returned
        else:
            source[key] = overrides[key]
    return source

def process_message(params):
    rng = random.SystemRandom()
    defaults = {
        "stationid": rng.randint(1,65535),
        "gn": {
            "gn_addr": "0:0:0:1",
            "geobc_fwd_alg": 0,
        },
        "denm": {
            "forwarding": True,
            "forceactionid": False,
            "autoupdate": False,
        },
    }


    defaults = _dict_update(defaults, params)

    envname = "env" + str(defaults["id"])

    logging.debug("Processing received JSON:\n%s", pprint.pformat(defaults))

    response = "304 Not modified"

    try:
        status_env(envname)
    except subprocess.CalledProcessError, e:
        if defaults['action'] == 'start':
            response = start_env(defaults, envname)
        elif defaults['action'] == 'stop':
            logging.info("Environment %s already stopped", envname)
    else:
        if defaults['action'] == 'start':
            logging.info("Environment %s already started", envname)
        elif defaults['action'] == 'stop':
            response = stop_env(envname, defaults["id"])
    return response
